### PR TITLE
vectorized c51 implementation based on vdqn

### PR DESCRIPTION
**Pull request by mniju**
vc51 based on c51 atari and vdqn atari

### CODE REVIEW

File: cleanrl/atari/v_c51_atari.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        )
+
+    def observation(self, observation):
+        return np.transpose(observation, axes=(2, 0, 1))
+
+def wrap_pytorch(env):
+    return ImageToPyTorch(env)
+
+# Reference: https://arxiv.org/pdf/1707.06887.pdf
+# https://github.com/ShangtongZhang/DeepRL/blob/master/deep_rl/agent/CategoricalDQN_agent.py
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:58]: This entire part is not used, right? Please remove them

File: cleanrl/atari/v_c51_atari.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            action = torch.argmax(q_values, 1)
+            random_action = torch.randint(0, envs.single_action_space.n, (len(x),), device=device)
+            random_action_flag = torch.rand(len(x), device=device) > epsilon
+            action = torch.where(random_action_flag, action, random_action)
+           
+        return action, pmfs[torch.arange(len(x)), action]
+
+def linear_schedule(start_e: float, end_e: float, duration: int, t: int):
+    slope =  (end_e - start_e) / duration
+    return max(slope * t + start_e, end_e)
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:233]: Please move the Agent class and functions at the top level of the script

File: cleanrl/atari/v_c51_atari.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    parser.add_argument('--train-frequency', type=int, default=4,
+                        help="the frequency of training")
+    args = parser.parse_args()
+    args.batch_size = int(args.num_envs * args.num_steps)
+    args.minibatch_size = int(args.batch_size // args.num_minibatches)
+    if not args.seed:
+        args.seed = int(time.time())
+    return args
+
+# TRY NOT TO MODIFY: setup the environment
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:143]: use a `if __name__ == "__main__": here and indent the code below

File: cleanrl/atari/v_c51_atari.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+for update in range(1, num_updates + 1):
+    # ROLLOUTS
+    epsilon = linear_schedule(args.start_e, args.end_e, args.exploration_fraction*args.total_timesteps, global_step)
+    writer.add_scalar("charts/epsilon", epsilon, global_step)
+    for step in range(0, args.num_steps):
+        global_step += 1 * args.num_envs
+        obs[step] = next_obs
+        dones[step] = next_done
+        with torch.no_grad():
+            action, _ = q_network.get_action(envs, device, next_obs)
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:268]: Where is the epsilon greedy exploration? In VDQN, I had [these lines](https://github.com/vwxyzjn/cleanrl/pull/59/files#diff-bfb142a3850419f93ee126b7e8ec527b4bca61221809062a7030e8f147685bd5R199-R206)

```python
            # ALGO LOGIC: action logic
            with torch.no_grad():
                logits = q_network.forward(next_obs)
                action = torch.argmax(logits, dim=1)
                random_action = torch.randint(0, envs.single_action_space.n, (envs.num_envs,), device=device)
                random_action_flag = torch.rand(envs.num_envs, device=device) > epsilon
                action = torch.where(random_action_flag, action, random_action)
            actions[step] = action
```

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
Great work Niju! Would you mind tracking this run with W&B?

File: cleanrl/atari/v_c51_atari.py (hunk 4 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:268]: The epsilon greedy exploration is in the Agents get_action method.

```
def get_action(self, envs, device, x,action=None):
        logits = self.forward(x, device)
        # probability mass function for each action
        pmfs = torch.softmax(logits.view(len(x), envs.single_action_space.n, self.n_atoms), dim=2)
        q_values = (pmfs*self.atoms).sum(2)
        if action is None:
            action = torch.argmax(q_values, 1)
            random_action = torch.randint(0, envs.single_action_space.n, (len(x),), device=device)
            random_action_flag = torch.rand(len(x), device=device) > epsilon
            action = torch.where(random_action_flag, action, random_action)

        return action, pmfs[torch.arange(len(x)), action]
```

File: cleanrl/atari/v_c51_atari.py (hunk 1 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:58]: Updated

File: cleanrl/atari/v_c51_atari.py (hunk 2 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:233]: Updated

File: cleanrl/atari/v_c51_atari.py (hunk 3 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:143]: Updated

File: cleanrl/atari/v_c51_atari.py (hunk 1 above)
Comment (vwxyzjn) [reply cleanrl/atari/v_c51_atari.py:58]: Thank you. Could you push the code please?

**Comment by mniju** [comment]
> Great work Niju! Would you mind tracking this run with W&B?

I have added a run here: https://wandb.ai/niju/VC51

**Comment by vwxyzjn** [comment]
Hey Niju, sorry for getting back to you late. Thanks for the re-factoring, the code looks a lot better.


Here is a file diff between vdqn and vc51: https://www.diffchecker.com/SDY62J2l. Would you mind refactoring it a bit more, minimizing the difference between vdqn and vc51?

**Comment by mniju** [comment]
> Hey Niju, sorry for getting back to you late. Thanks for the re-factoring, the code looks a lot better.
> 
> Here is a file diff between vdqn and vc51: https://www.diffchecker.com/SDY62J2l. Would you mind refactoring it a bit more, minimizing the difference between vdqn and vc51?

Hi Costa , I made some cleanup to minimize difference between vdqn and vc51. Please let me know if further refactoring is required, Thanks

File: cleanrl/atari/v_c51_atari.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    parser = argparse.ArgumentParser(description='Vectorized-C51 agent')
+    # Common arguments
+    parser.add_argument('--exp-name', type=str, default=os.path.basename(__file__).rstrip(".py"),
+                        help='the name of this experiment')
+    parser.add_argument('--gym-id', type=str, default="BreakoutNoFrameskip-v4",
+                        help='the id of the gym environment')
+    parser.add_argument('--learning-rate', type=float, default=1e-4,
+                        help='the learning rate of the optimizer')
+    parser.add_argument('--seed', type=int, default=2,
+                        help='seed of the experiment')
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:35]: 1. Would you mind removing the extra spaces in the argparse from 
```
	    parser.add_argument('--exp-name', type=str, default=os.path.basename(__file__).rstrip(".py"),
                        help='the name of this experiment')
```

File: cleanrl/atari/v_c51_atari.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    FireResetEnv,
+    MaxAndSkipEnv,
+    NoopResetEnv,
+)
+from torch.utils.tensorboard import SummaryWriter
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Vectorized-C51 agent')
+    # Common arguments
+    parser.add_argument('--exp-name', type=str, default=os.path.basename(__file__).rstrip(".py"),
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:28]: add `# fmt: off` as a new line before `parser.add....`

File: cleanrl/atari/v_c51_atari.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    parser.add_argument('--start-e', type=float, default=1.,
+                        help="the starting epsilon for exploration")
+    parser.add_argument('--end-e', type=float, default=0.01,
+                        help="the ending epsilon for exploration")
+    parser.add_argument('--exploration-fraction', type=float, default=0.10,
+                        help="the fraction of `total-timesteps` it takes from start-e to go end-e")
+
+    args = parser.parse_args()
+    args.batch_size = int(args.num_envs * args.num_steps)
+    args.minibatch_size = int(args.batch_size // args.num_minibatches)
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:81]: add a `# fmt:on` here

File: cleanrl/atari/v_c51_atari.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    return args
+
+# TRY NOT TO MODIFY: setup the environment
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:84]: remove this

File: cleanrl/atari/v_c51_atari.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+# env setup
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:86]: remove this too

File: cleanrl/atari/v_c51_atari.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        env.seed(seed)
+        env.action_space.seed(seed)
+        env.observation_space.seed(seed)
+        return env
+
+    return thunk
+
+
+class QNetwork(nn.Module):
+    def __init__(self, envs, device, frames=4, n_atoms=51, v_min=-10, v_max=10):
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:112]: remove `device` as part of the arguments

File: cleanrl/atari/v_c51_atari.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+            nn.Flatten(),
+            nn.Linear(3136, 512),
+            nn.ReLU(),
+            nn.Linear(512, envs.single_action_space.n * n_atoms)
+        )
+
+    def forward(self, x, device):
+        return self.network(x/255.0)
+
+    def get_action(self, envs, device, x,action=None):
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:132]: remove device in the function argument

File: cleanrl/atari/v_c51_atari.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+            '\n'.join([f"|{key}|{value}|" for key, value in vars(args).items()])))
+    if args.track:
+        import wandb
+        wandb.init(project=args.wandb_project_name, 
+                   entity=args.wandb_entity,
+                   sync_tensorboard=True,
+                   config=vars(args),
+                   name=run_name,
+                   monitor_gym=True,
+                   save_code=True)
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:163]: replace this part of the code with 

```
    args = parse_args()
    run_name = f"{args.gym_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
    if args.track:
        import wandb

        wandb.init(
            project=args.wandb_project_name,
            entity=args.wandb_entity,
            sync_tensorboard=True,
            config=vars(args),
            name=run_name,
            monitor_gym=True,
            save_code=True,
        )
    writer = SummaryWriter(f"runs/{run_name}")
    writer.add_text(
        "hyperparameters",
        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
    )
```

File: cleanrl/atari/v_c51_atari.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+    envs = gym.vector.SyncVectorEnv(
+        [make_env(args.gym_id, args.seed + i, i, args.capture_video, run_name) for i in range(args.num_envs)]
+    )
+    assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+
+    # ALGO LOGIC: initialize agent here:
+    q_network = QNetwork(envs, device, n_atoms=args.n_atoms, v_min=args.v_min, v_max=args.v_max).to(device)
+    target_network = QNetwork(envs, device, n_atoms=args.n_atoms, v_min=args.v_min, v_max=args.v_max).to(device)
+    target_network.load_state_dict(q_network.state_dict())
+    optimizer = optim.Adam(q_network.parameters(), lr=args.learning_rate, eps=0.01/args.batch_size)
```
Comment (vwxyzjn) [inline cleanrl/atari/v_c51_atari.py:183]: is modifying `eps` necessary?

**Review by vwxyzjn — CHANGES_REQUESTED** [review]
Thank you for the change. A few more things as pointed out. See the file diff here https://www.diffchecker.com/QXNTmZmN

File: cleanrl/atari/v_c51_atari.py (hunk 6 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:28]: Sure , i would like to know  the purpose of this comment # fmt: off , # fmt: on

Comment (vwxyzjn) [reply cleanrl/atari/v_c51_atari.py:28]: This is for formatting. After doing it try doing `black vc51_.py` (Auto-formatting). Also would you mind renaming the file to `vc51_atari.py`? Sorry for the trouble.

Comment (vwxyzjn) [reply cleanrl/atari/v_c51_atari.py:28]: I think the full command I use is 

```
isort . --skip wandb
autoflake -r --exclude wandb --in-place --remove-unused-variables --remove-all-unused-imports .
black -l 127 --exclude wandb .
```

Also, would you mind putting this file under a folder called `vectorized_architecture`?

Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:28]: Costa , I will make the required  changes. Keep me posted of all the required changes

File: cleanrl/atari/v_c51_atari.py (hunk 13 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py:183]: I got it from c51_atari.py file. Can remove it if required

https://github.com/vwxyzjn/cleanrl/blob/d1002bd80cecda4e7251460a1db1c832943578e8/cleanrl/c51_atari.py#L491

File: cleanrl/atari/v_c51_atari.py (hunk 6 above)
Comment (mniju) [reply cleanrl/atari/v_c51_atari.py]: Pushed the formatted code

File: cleanrl/atari/v_c51_atari.py (hunk 13 above)
Comment (vwxyzjn) [reply cleanrl/atari/v_c51_atari.py]: Oh interesting. I forgot about this. Then keep it as it is please

**Comment by vwxyzjn** [comment]
Thank you!

**Pull request merged**